src/main/java/org/jabref/gui/maintable/MainTableHeaderContextMenu.java

```diff
@@ -0,0 +1,184 @@
+     */
+    public void show(boolean show) {
+        // TODO: 20/10/2022 unknown bug where issue does not show unless parameter is passed through this method.
+        mainTable.setOnContextMenuRequested(event -> {
+            // Display the menu if header is clicked, otherwise, remove from display.
+            if (!(event.getTarget() instanceof StackPane) && show) {
+                this.show(mainTable, event.getScreenX(), event.getScreenY());
+            } else if (this.isShowing()) {
+                this.hide();
+            }
+            event.consume();
+        });
+    }
+
+    /**
+     * Constructs the items for the list and places them in the menu.
+     */
+    private void constructItems(MainTable mainTable) {
+        // Reset the right-click menu
+        this.getItems().clear();
+
+        // Populate the menu with the commonly used fields
+        for (TableColumn<BibEntryTableViewModel, ?> tableColumn:commonColumns()) {
+            RadioMenuItem itemToAdd = createMenuItem(tableColumn);
+            this.getItems().add(itemToAdd);
+        }
+
+        SeparatorMenuItem separator = new SeparatorMenuItem();
+        this.getItems().add(separator);
+
+        // Append to the menu the current remaining columns in the table.
+        for (TableColumn<BibEntryTableViewModel, ?> column :mainTable.getColumns()
+        ) {
+            // Append only if the column has not already been added (a common column)
+            if (!isACommonColumn((MainTableColumn) column)) {
+                RadioMenuItem itemToAdd = createMenuItem(column);
+                this.getItems().add(itemToAdd);
+            }
+        }
+    }
+
+    /**
+     * Creates an item for the menu constructed with the name/visibility of the table column.
+     *
+     */
+    @SuppressWarnings("rawtypes")
+    private RadioMenuItem createMenuItem(TableColumn<BibEntryTableViewModel, ?> column) {
+        // Construct initial menuItem
+        MainTableColumn tableColumn = (MainTableColumn) column;
+        String itemName = tableColumn.getDisplayName();
+
+        RadioMenuItem returnItem = new RadioMenuItem(itemName);
+
+        // Flag item as selected if the item is already in the main table.
+        returnItem.setSelected(isInMainTable(tableColumn));
+
+        // Set action to toggle visibility from main table when item is clicked
+        returnItem.setOnAction(event -> {
+            if (isInMainTable(tableColumn)) {
+                removeColumn(tableColumn);
+            } else {
+                addColumn(tableColumn);
+            }
+        });
+
+        return returnItem;
+    }
+
+    /**
+     * Adds the column into the MainTable for display.
+     */
+    @SuppressWarnings("rawtypes")
+    private void addColumn(MainTableColumn tableColumn) {
+        // Do not add duplicate if table column is already within the table.
+        if (!isInMainTable(tableColumn)) {
+            mainTable.getColumns().add(tableColumn);
```

Comment:
What happens if the user first removes a column, and then immediately readds it (e.g. he realizes removing it was a mistake). In this case, the column order should be the same as it was before. If I read the code correctly, then currently the column would be appended at the end instead of at its original position.

Reply:
I've made some more changes here. When the user goes to click to remove the column, it obtains the current index in the MainTable and stores that number inside the Menu Item. If he wants to immediately place it back, it will be placed back in that exact position using that index.
If the items next to it are removed, it will still be placed in the position of the given index - not sure if that will be confusing to the end user or not, happy to hear your thoughts.